Use ILogger from Microsoft instead of our own abstraction

amoerie opened the pull request:
Fixes #1314

@gofal
I'm making this pull request to showcase what it would look like if we dropped our own ILogger abstraction.
~~Obviously, do not merge this. None of the samples/docs have been updated~~, but all the code compiles and the tests are green.

Since logging is not really the core of this library, I think it might be okay to remove our custom abstraction in favor of Microsoft's.  Every logging library provides adapters that plug into this abstraction.

Benefits that I can see so far:
- No need to make adapters for our ILogger interface -> some logging library. We currently maintain a package for NLog and Serilog, but that would become redundant then.
- Microsoft.Extensions.Logging is a lot more performant when certain log levels are disabled. I will post some benchmarks soon.
- We can remove a bunch of code that is not relevant to DICOM at all: our NLog project, some utilities like NameFormatToPositionalFormat, etc.
- Access to a new log level we didn't have yet: Trace
- Access to high performance APIs for logging (see https://docs.microsoft.com/en-us/dotnet/core/extensions/high-performance-logging and https://docs.microsoft.com/en-us/dotnet/core/extensions/logger-message-generator )
- Better integration with .NET Core. Logging will now work out of the box without having to configure a LogManager.
- Better tooling support. JetBrains Rider and Visual Studio now highlight usages of ILogger with intellisense for structured logging

Downsides/risks that I can see:
- Breaking change. Everyone that updates would have to revisit their logging configuration. Some "simple" consumers in .NET Core might have to do nothing if they already configured Microsoft.Extensions.Logging.
- Microsoft.Extensions.Logging works against .NET Standard 2.0, but we will have to hope they keep it that way for another long time, otherwise we will be unable to upgrade this dependency

#### Checklist
- [x] The pull request branch is in sync with latest commit on the *fo-dicom/development* branch
- [x] I have updated API documentation
- [x] I have included unit tests
- [x] I have updated the change log
- [x] I am listed in the CONTRIBUTORS file

#### Changes proposed in this pull request:
- Remove our own ILogger, Logger and LogManager
- Remove FO-DICOM.NLog
- Use ILogger from Microsoft.Extensions.Logging everywhere directly

amoerie committed:
Use ILogger from Microsoft instead of our own

amoerie committed:
Fix FO-DICOM.AspNetCore

amoerie committed:
Fix a few warnings

amoerie committed:
Remove unneeded dependency

amoerie committed:
Remove .NET Core 2.1 from github yml config

amoerie committed:
Merge branch 'development' of github.com:fo-dicom/fo-dicom into GH-1314

amoerie committed:
Don't collect logs by default, only when enabled

amoerie committed:
Remove reference to our old NLog project in tests

amoerie committed:
Apply LogDimseDatasets correctly

amoerie committed:
Reinstate normal benchmarks

amoerie commented:
Benchmarks:

``` ini

BenchmarkDotNet=v0.12.0, OS=Windows 10.0.22000
Intel Core i7-8700K CPU 3.70GHz (Coffee Lake), 1 CPU, 12 logical and 6 physical cores
.NET Core SDK=6.0.102
  [Host]     : .NET Core 3.1.22 (CoreCLR 4.700.21.56803, CoreFX 4.700.21.57101), X64 RyuJIT
  Job-XBBRLV : .NET Core 3.1.22 (CoreCLR 4.700.21.56803, CoreFX 4.700.21.57101), X64 RyuJIT

InvocationCount=128  MaxIterationCount=25  MaxWarmupIterationCount=10  
UnrollFactor=16  

```
|            Type |            Method | Logging  |  Mean |     Error |    StdDev |   Gen 0 |  Gen 1 | Gen 2 | Allocated |
|----------:|------------------ |----------:|---------:|----------:|----------:|--------:|-------:|------:|----------:|
| Old |  SendEchoToServer | Disabled | 1.266 ms | 0.142 ms | 0.1905 ms | 15.6250 |      - |     - |  96.29 KB |
| New |  SendEchoToServer |   Disabled | 0.968 ms |  0.053 ms |  0.069 ms |  7.8125 |     - |     - |  80.94 KB |
| Old |  SendStoreToServer | Disabled | 3.512 ms | 0.245 ms | 0.3185 ms | 23.4375 | 7.8125 |     - | 142.72 KB |
| New | SendStoreToServer | Disabled | 2.993 ms | 0.213 ms | 0.269 ms | 15.6250 |     - |     - | 126.24 KB |
| Old |  SendEchoToServer |  Console | 9.023 ms | 0.0245 ms | 0.0205 ms | 23.4375 | 7.8125 |     - | 178.92 KB |
| New |  SendEchoToServer |  Console | 8.987 ms | 0.1127 ms | 0.0745 ms | 31.2500 |       - |     - |  190.7 KB |
| Old | SendStoreToServer | Console |  16.324 ms | 0.2583 ms | 0.2016 ms | 62.5000 | 7.8125 |     - |  415.4 KB |
| New | SendStoreToServer | Console | 15.940 ms | 0.1062 ms | 0.0768 ms | 85.9375 | 15.6250 |     - | 541.69 KB |

Some preliminary conclusions (I think our usages of ILogger could be optimized some more):

- ILogger is definitely faster and allocates less if logging / the log level is disabled
- ILogger, in the specific case of Console logging at least, seems to be similar in speed but higher in allocations. But this is not super relevant, because I would expect nobody runs their applications with console logging in production. I think people typically log to files or to remote log storage mechanisms such as Seq.

amoerie committed:
Merge branch 'development' into GH-1314

gofal commented:
most of our production code is using logging to console. we are running most of our applications in docker containers and therefore are just simple console applications. the output of the console within docker is then redirected to some logs outside docker.

gofal commented:
This looks very interesting to me. Although you are right, that would be a significant breaking change.

amoerie commented:
> most of our production code is using logging to console. we are running most of our applications in docker containers and therefore are just simple console applications. the output of the console within docker is then redirected to some logs outside docker.

That's a very good point! 

> That would be a significant breaking change.

Maybe it doesn't have to be. We could keep the old ILogger for now, mark it as obsolete and provide an adapter from our ILogger -> Microsoft ILogger that is configured by default.

> This looks very interesting to me

If we want to pursue this, I could clean up this PR and update the docs.
I also saw in another discussion that we haven't migrated our Serilog adapter to v5 yet. If we do this PR instead, that would be redundant.
If you think about it, we also don't have our own abstraction for a dependency container because the one provided by Microsoft (IServiceProvider) is universal. The same applies to logging, imho.

mrbean-bremen commented:
> Maybe it doesn't have to be. We could keep the old ILogger for now, mark it as obsolete and provide an adapter from our ILogger -> Microsoft ILogger that is configured by default.

That sounds good to me - I also like the idea of using the MS logger, and maybe the old logger can be removed in fo-dicom 6.

amoerie committed:
Merge development into GH-1314

amoerie committed:
Reintroduce an obsolete version of the old ILogger

amoerie committed:
Log exceptions using the recommended API

amoerie committed:
Add adapter for old legacy ILogger

amoerie committed:
Merge development into GH-1314

amoerie committed:
Adapt new advanced DICOM client API to Microsoft.Extensions.Logging

amoerie committed:
Update README

amoerie committed:
Don't mention fo-dicom.NLog anymore in the README

amoerie committed:
Update README with usage instructions for logging

amoerie committed:
Update logging docs

amoerie committed:
Cleanup

- Move obsolete APIs to a separate namespace to avoid namespace conflicts
- Make list of used logging categories explicit

amoerie committed:
Document namespace FellowOakDicom.Log.Obsolete

amoerie commented:
So, I pursued this a little further:

- Resolved all conflicts with development
- Adapted the documentation (both README and the DocFx page about logging)
- Provided a backwards compatibility path for people who want to keep using `ILogManager` and documented how to to this

The new logging documentation now looks like this:

## Introduction
*fo-dicom* relies on Microsoft.Extensions.Logging for its logging purposes.
Any logging configuration that is configured there, will automatically be used.

For example, when codecs are loaded, their names are written through the logger at a debug level.

## Default configuration
By default, *fo-dicom* will make use of `Microsoft.Extensions.Logging` for all of its logging.

More specifically, *fo-dicom* injects loggers of type `Microsoft.Extensions.Logging.ILogger` where necessary.

## Default logging libraries
To use NLog, Serilog or other logging libraries, simply configure them as you would for `Microsoft.Extensions.Logging` and they will be used automatically by *fo-dicom*.

## Backwards compatibility configuration
In the past, *fo-dicom* used custom abstractions for logging called `FellowOakDicom.Log.ILogManager` and `FellowOakDicom.Log.ILogger`
Using these historical abstractions is still supported, but they have moved to the `FellowOakDicom.Log.Obsolete` namespace and are marked obsolete.

It is possible to change the log manager with the following code:
```
new DicomSetupBuilder()
    .RegisterServices(s => s.AddLogManager<MyLogManager>())
    .Build();
```

__Warning__: internally, this `MyLogManager` will be registered as an `ILoggerProvider` (through an adapter) in `Microsoft.Extensions.Logging`
If you use `Microsoft.Extensions.Logging` elsewhere in the same application, your `MyLogManager` will also appear there.

## Backwards compatibility logging libraries
NLog is officially supported via a pre-built library available on nuget: https://www.nuget.org/packages/fo-dicom.NLog/

To configure *NLog* integration with *fo-dicom*, simply reference the nuget package, and insert the following code
```
new DicomSetupBuilder()
    .RegisterServices(s => s.AddLogManager<NLogManager>())
    .Build();
```

For all other logging libraries, you will have to provide your own implementation of `ILogManager` and `ILogger`.

amoerie committed:
Cleanup merge artifacts

amoerie committed:
Merge branch 'development' into GH-1314

amoerie committed:
Merge branch 'development' into GH-1314

amoerie committed:
Fix logging usage in new code since last merge

amoerie committed:
Merge branch 'development' into GH-1314

amoerie committed:
Merge branch 'GH-1314' of github.com:fo-dicom/fo-dicom into GH-1314

amoerie committed:
Make ClientHandleNEventReport less flaky

amoerie committed:
Merge branch 'development' of github.com:fo-dicom/fo-dicom into GH-1314

amoerie committed:
Make DicomServerStoppedOnPort less flaky

amoerie committed:
Merge with development

amoerie committed:
Merge with development

amoerie committed:
Merge branch 'development' of github.com:fo-dicom/fo-dicom into GH-1314

gofal committed:
Merge branch 'development' into GH-1314

amoerie committed:
Merge branch 'development' of github.com:fo-dicom/fo-dicom into GH-1314

amoerie committed:
Merge development

amoerie commented:
Pinging @gofal @mrbean-bremen for a review, if possible :-)

mrbean-bremen commented:
Won't get to it today, but I will try to get a look later this week.

mrbean-bremen commented on FO-DICOM.Core/Log/LogCategories.cs:
```diff
@@ -0,0 +1,17 @@
+﻿// Copyright (c) 2012-2022 fo-dicom contributors.
```
2023?

mrbean-bremen commented on README.md:
```diff
@@ -49,35 +48,56 @@ Documentation, including API documentation, is available via GitHub pages:
-    s.AddLogManager<NLogManager>()   // or ...
-    
+In the past, Fellow Oak DICOM had a custom abstraction for logging: ILogger and ILogManager.
+For backwards compatibility purposes, this is still supported but not recommended for new applications.
```
Do we want to give some estimate as to when the deprecated stuff will be removed (certainly not too soon)? 
Maybe give some assurance that the old logger, while deprecated, will remain (unsupported) for a sufficiently long time.

I like the changes and think that this is the way to go, but I'm a bit concerned about the potential impact this has on users who use the old logging (and maybe wrote their own loggers based on `ILogManager`) _if_ they have to rewrite their logging code.  

Also maybe it would help to later add some adapter, so they at least don't have to rewrite all logging calls in their code (as you had to here). Though I would wait for user feedback before doing anything like this.

gofal replied:
Agree. For that reason, I would prefer not to move the old ILogManager into the `Obsolete` namespace. Because this would really mean to touch every single file where logging is used if updating to version 5.1. 

What would be the downside of having the old logger classes in namespace `FellowOakDicom.Log.ILogger`? In our code the microsoft logge is the default, and in the backwards-compatibility-classes we would have to access the class with the whole name including namespace. Right?

amoerie replied:
The downside is mostly that ILogger exists both in FellowOakDicom.Log and Microsoft.Extensions.Logging, but you're right that this is solvable by using the full namespace where needed. I've updated the code to no longer use the "Obsolete" namespace to reduce friction for people who upgrade.

amoerie replied:
> Do we want to give some estimate as to when the deprecated stuff will be removed (certainly not too soon)?
> Maybe give some assurance that the old logger, while deprecated, will remain (unsupported) for a sufficiently long time.

We can definitely keep it around for one or more years, there is no hurry to remove it since it doesn't really bother us maintenance wise. However, I would refrain from making any promises. 

> Also maybe it would help to later add some adapter, so they at least don't have to rewrite all logging calls in their code (as you had to here). Though I would wait for user feedback before doing anything like this.

Actually the adapter already exists. If you wrap your custom ILogger in a FellowOakDicomLogger you can register it directly into the Microsoft.Extensions.Logging setup and skip the obsolete LogManager.

amoerie replied:
@gofal if you have the time, please can you give this another look :-)

amoerie committed:
Merge

amoerie committed:
Move old logging code out of Obsolete namespace

amoerie committed:
Remove duplicate packages

gofal committed:
Merge branch 'development' into GH-1314

gofal committed:
fix build issues after merging

gofal committed:
fix assertion in unittest
